### ISSUE TITLE
Fusion of torch.nn.functional.gelu does not work if gradients are required

### ISSUE THREAD

**Issue by kevinstephano** [csarofeen/pytorch#427]
## 🐛 Bug

Can't fuse with native GELU in pytorch during training.

## To Reproduce

Just forward:

```
import torch
import os

torch._C._jit_set_nvfuser_enabled(True)
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)
torch._C._jit_override_can_fuse_on_cpu(False)
torch._C._jit_override_can_fuse_on_gpu(False)
torch._C._jit_set_bailout_depth(20)

os.environ['PYTORCH_CUDA_FUSER_DEBUG'] = '1'

@torch.jit.script
def bias_gelu(bias, y):                                                                                                                                                                 
    x = bias + y
    return torch.nn.functional.gelu(x)

input_shape = [6, 512, 4096]
grad_shape = [6, 512, 4096]
bias_shape = [6, 512, 4096]

inputs = torch.randn(input_shape, device='cuda', dtype=torch.float16, requires_grad=False)
bias = torch.randn(bias_shape, device='cuda', dtype=torch.float16, requires_grad=False)

for idx in range(0, 5) :
    if idx == 1 :
        print(bias_gelu.graph_for(bias, inputs))
    outputs = bias_gelu(inputs, bias)
```

With backward:

```
import torch
import os

torch._C._jit_set_nvfuser_enabled(True)
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)
torch._C._jit_override_can_fuse_on_cpu(False)
torch._C._jit_override_can_fuse_on_gpu(False)
torch._C._jit_set_bailout_depth(20)

os.environ['PYTORCH_CUDA_FUSER_DEBUG'] = '1'

@torch.jit.script
def bias_gelu(bias, y):                                                                                                                                                                 
    x = bias + y
    return torch.nn.functional.gelu(x)

input_shape = [6, 512, 4096]
grad_shape = [6, 512, 4096]
bias_shape = [4096]

inputs = torch.randn(input_shape, device='cuda', dtype=torch.float16, requires_grad=True)
grads = torch.randn(grad_shape, device='cuda', dtype=torch.float16, requires_grad=True)
bias = torch.randn(bias_shape, device='cuda', dtype=torch.float16, requires_grad=True)

for idx in range(0, 5) :
    if idx == 1 :
        print(bias_gelu.graph_for(bias, inputs))
    outputs = bias_gelu(inputs, bias)
    outputs.backward(grads)
```

**Comment by kevinstephano** [comment]
Note, torch.nn.functional.gelu is also not preferred to be used because it is slow due to this issue filed on the core side: http://nvbugs/200607482.

**Comment by jjsjann123** [comment]
few issues I found so far:
1. `gelu` is not autodiff~able: https://github.com/pytorch/pytorch/blob/61623430d32c081124067fc4abf33bd4a023af90/torch/csrc/jit/runtime/symbolic_script.cpp#L1494

~2. `profile` node right now could block merging autodiff nodes, since we changed the timing of when profiling nodes are removed: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp#L112-L127~
NVM the second one, `profile` node is fine as it's getting merged to autodiff subgraph as well. :stuck_out_tongue_closed_eyes:

**Comment by kevinstephano** [comment]
Change to `kernel_resource_strings.h`:

```
--- a/torch/csrc/jit/codegen/cuda/kernel_resource_strings.h
+++ b/torch/csrc/jit/codegen/cuda/kernel_resource_strings.h
@@ -159,7 +159,7 @@ __device__ float frac(const float x) {
   return x - truncf(x);
 }
 __device__ float gelu(const float x) {
-  return x * normcdf(x);
+  return x * 0.5 * (1.0 + erff(x * 0.70710678));
 }
 __device__ float reciprocal(const float x) {
   return 1.f / x;
```

Change to  `jit\runtime\symbolic_scripts.cpp`:

```
        def gelu(self):
            result = torch.gelu(self)
            def backward(grad_output):
                return 0.5 * (1. + torch.erf(self * 0.70710678)) + 0.3989423 * self * torch.exp(-0.5 * self * self) * grad_output

            return result, backward
```

**Comment by kevinstephano** [comment]
I created this PR on upstream: https://github.com/pytorch/pytorch/pull/46785

**Comment by csarofeen** [comment]
I believe this is fixed, reopen if there's still an issue.
